### PR TITLE
[language] loader cleanup

### PR DESCRIPTION
**Pull request by dariorussi**
Refactor code a bit to organize it better for readability.
Added some commments.
Cleaned up `Loader` and `Resolver` API.
There is more coming but this is a good change

### CODE REVIEW

**Comment by dariorussi** [comment]
this is not a big change but moving code around is reporting a much bigger number of line changed that it really is

File: language/move-vm/runtime/src/loader.rs [hunk 1]
```diff
@@ -386,10 +371,105 @@ impl Loader {
...
+    // Script verification steps.
+    // See `verify_module()` for module verification steps.
+    fn verify_script(&self, script: &CompiledScript) -> VMResult<()> {
+        DuplicationChecker::verify_script(&script)?;
+        SignatureChecker::verify_script(&script)?;
+        InstructionConsistency::verify_script(&script)?;
+        constants::verify_script(&script)?;
+        CodeUnitVerifier::verify_script(&script)?;
+        verify_main_signature(&script)
+    }
```
Comment (tnowacki) [inline language/move-vm/runtime/src/loader.rs:422]: This looks like the same code in verifier.rs? Why the duplication?

Comment (dariorussi) [reply language/move-vm/runtime/src/loader.rs:422]: I want the runtime to have its own concept of verification and I do not want to use a common function.
Both `verify_module` and `verify_script` are implemented by the runtime.
Maybe I could point that out in the comment but it really seem too much information.
Bottom line I do not want the runtime to share anything with anybody when it comes to core logic.
Of course that has a lot of not so nice side effects together with a lot of very nice, and truth is, I don't know what's right.
But I really do not want to share for now... I want to be really selfish here... ;) 
thoughts?

**Comment by dariorussi** [comment]
there is one more PR coming for the loader and built on top of this. 
If people think is better I am happy to merge it here.
Also I am fine to wait going in if people have bad merges, please let me know. Happy to wait.
I am cleaning up the loader on top of this and it's all internal API so I don't have to go in right away if I am disruptive to others

**Comment by vgao1996** [comment]
@dariorussi Sorry tried to land the wrong PR... Will get the flow canceled

**Comment by vgao1996** [comment]
Still I'm surprised bors-libra permitted this...

**Comment by dariorussi** [comment]
> Still I'm surprised bors-libra permitted this...

this is hilarious!!!!

File: language/move-vm/runtime/src/loader.rs [hunk 2]
```diff
@@ -690,32 +742,69 @@ impl<'a> Resolver<'a> {
...
             BinaryType::Script(script) => script.function_instantiation_at(idx.0),
+        };
+        let mut instantiation = vec![];
+        for ty in &func_inst.instantiation {
+            instantiation.push(ty.subst(type_params)?);
         }
+        Ok(instantiation)
+    }
+
+    pub(crate) fn generic_function_arity(
```
Comment (vgao1996) [inline language/move-vm/runtime/src/loader.rs:792]: I found this name a bit confusing. Upon first glance, "arity" could mean a few things to me:
1. The total number of parameters, including type parameters and value parameters
2. The number of value parameters.
3. The number of type parameters.
and it was only clear until I read the body of the function.

I'd personally prefer a more straightforward name including sth like "num_type_params".

Comment (dariorussi) [reply language/move-vm/runtime/src/loader.rs:792]: went for `type_params_count`

**Comment by dariorussi** [comment]
some feedback

**Pull request closed without merging**